lgc/builder/llpcYCbCrAddressHandler.cpp

```diff
@@ -0,0 +1,190 @@
+ *  copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+/**
+***********************************************************************************************************************
+* @file  llpcYCbCrAddressHandler.cpp
+* @brief LLPC source file: Implementation of LLPC class YCbCrAddressHandler
+***********************************************************************************************************************
+*/
+#include "llpcYCbCrAddressHandler.h"
+#include "llpcGfxRegHandler.h"
+#include "llpcInternal.h"
+
+#include "llvm/IR/Intrinsics.h"
+#include "llvm/IR/IntrinsicsAMDGPU.h"
+
+using namespace lgc;
+using namespace llvm;
+
+// =====================================================================================================================
+// Generate base address of each image plane
+void YCbCrAddressHandler::GenBaseAddress(
+    uint32_t planeNum) // The count of planes
+{
+    if (planeNum > 0)
+    {
+        // m_PlaneBaseAddresses[0] is the same as original base addressed as passed in SRD
+        m_PlaneBaseAddresses.push_back(m_pRegHelper->GetReg(SqRsrcRegs::BaseAddress));
+        if (planeNum > 1)
+        {
+            // m_PlaneBaseAddresses[1] = m_PlaneBaseAddresses[0] + addr_256B(m_pPitchY * m_pHeightY)
+            m_PlaneBaseAddresses.push_back(m_pBuilder->CreateAdd(m_PlaneBaseAddresses[0],
+                                           m_pBuilder->CreateLShr(m_pBuilder->CreateMul(m_pPitchY, m_pHeightY),
+                                           m_pBuilder->getInt32(8))));
+            if (planeNum > 2)
+            {
+                // m_PlaneBaseAddresses[2] = m_PlaneBaseAddresses[1] + addr_256B(m_pPitchCb * m_pHeightCb)
+                m_PlaneBaseAddresses.push_back(m_pBuilder->CreateAdd(m_PlaneBaseAddresses[1],
+                                               m_pBuilder->CreateLShr(m_pBuilder->CreateMul(m_pPitchCb, m_pHeightCb),
+                                               m_pBuilder->getInt32(8))));
+            }
+        }
+    }
+}
+
+// =====================================================================================================================
+// Power2align operation
+Value* YCbCrAddressHandler::Power2Align(
+    Value*   pX,    // [in] Value needs to be aligned
+    uint32_t align) // Align base
+{
+    uint32_t alignM1 = align - 1;
```

Comment:
Can we assert that `align` is a power of 2?

Reply:
Added an assert.